htsget-search/src/storage/url.rs

```diff
@@ -0,0 +1,494 @@
+/// A storage struct which derives data from HTTP URLs.
+#[derive(Debug, Clone)]
+pub struct UrlStorage {
+  client: Client,
+  url: Url,
+  response_scheme: Scheme,
+  forward_headers: bool,
+}
+
+impl UrlStorage {
+  /// Construct a new UrlStorage.
+  pub fn new(client: Client, url: Url, response_scheme: Scheme, forward_headers: bool) -> Self {
+    Self {
+      client,
+      url,
+      response_scheme,
+      forward_headers,
+    }
+  }
+
+  fn map_err<K: Display>(key: K, error: Error) -> StorageError {
+    match error.status() {
+      None => KeyNotFound(format!("{}", key)),
+      Some(status) => KeyNotFound(format!("for {} with status: {}", key, status)),
+    }
+  }
+
+  fn apply_headers(builder: RequestBuilder, headers: &HeaderMap) -> RequestBuilder {
+    headers
+      .iter()
+      .fold(builder, |builder, (key, value)| builder.header(key, value))
+  }
+
+  /// Get a url from the key.
+  pub fn get_url_from_key<K: AsRef<str> + Send>(&self, key: K) -> Result<Url> {
+    self
+      .url
+      .join(key.as_ref())
+      .map_err(|err| UrlParseError(err.to_string()))
+  }
+
+  /// Construct and send a request
+  pub async fn send_request<K: AsRef<str> + Send>(
+    &self,
+    key: K,
+    headers: &HeaderMap,
+    method: Method,
+  ) -> Result<reqwest::Response> {
+    let key = key.as_ref();
+    let url = self.get_url_from_key(key)?;
+    let url_key = url.to_string();
+
+    let builder = self.client.request(method, url);
+    let builder = Self::apply_headers(builder, headers);
+
+    builder
+      .send()
+      .await
+      .map_err(|err| Self::map_err(url_key, err))
+  }
+
+  /// Construct and send a request
+  pub fn format_url<K: AsRef<str> + Send>(
+    &self,
+    key: K,
+    options: RangeUrlOptions<'_>,
+  ) -> Result<HtsGetUrl> {
+    let mut url = self.get_url_from_key(key)?;
+
+    url
+      .set_scheme(&self.response_scheme.to_string())
+      .map_err(|_| {
+        InternalError("failed to set scheme when formatting response url".to_string())
+      })?;
+
+    let mut url = HtsGetUrl::new(url);
+    if self.forward_headers {
+      url = url.with_headers(
+        options
+          .response_headers()
+          .try_into()
+          .map_err(|err: error::Error| StorageError::InvalidInput(err.to_string()))?,
+      )
+    }
+
+    Ok(options.apply(url))
+  }
+
+  /// Get the head from the key.
+  pub async fn head_key<K: AsRef<str> + Send>(
+    &self,
+    key: K,
+    headers: &HeaderMap,
+  ) -> Result<reqwest::Response> {
+    self.send_request(key, headers, Method::HEAD).await
+  }
+
+  /// Get the key.
+  pub async fn get_key<K: AsRef<str> + Send>(
+    &self,
+    key: K,
+    headers: &HeaderMap,
+  ) -> Result<reqwest::Response> {
+    self.send_request(key, headers, Method::GET).await
+  }
+}
+
+#[async_trait]
+impl Storage for UrlStorage {
+  // There might be a nicer way to express this type.
+  type Streamable = StreamReader<Pin<Box<dyn Stream<Item = Result<Bytes>> + Send + Sync>>, Bytes>;
```

Comment:
Yikes, I recall there were some streaming RFCs in Tokio (and/or std::), perhaps worth revisiting and see if the state of the art is a bit less involved?